tracer/src/Datadog.Trace/AppSec/ControllerContextExtensions.Framework.cs

```diff
@@ -28,43 +28,55 @@ internal static void MonitorBodyAndPathParams(this IControllerContext controller
-                    {
-                        pathParamsDic[item.Key] = item.Value;
-                    }
-                    else
+                    scope = SharedItems.TryPeekScope(context, peekScopeKey);
+
+                    var bodyDic = new Dictionary<string, object>(parameters.Count);
+                    pathParamsDic = new Dictionary<string, object>(parameters.Count);
+                    foreach (var item in parameters)
                     {
-                        bodyDic[item.Key] = item.Value;
+                        if (controllerContext.RouteData.Values.ContainsKey(item.Key))
+                        {
+                            pathParamsDic[item.Key] = item.Value;
+                        }
+                        else
+                        {
+                            // We exclude the query string params
+                            if (!context.Request.QueryString.AllKeys.Contains(item.Key))
+                            {
+                                bodyDic[item.Key] = item.Value;
+                            }
+                        }
                     }
+
+                    requestBody = ObjectExtractor.Extract(bodyDic);
                 }
 
-                var securityTransport = new Coordinator.SecurityCoordinator(security, context, scope.Span);
-                if (!securityTransport.IsBlocked)
+                if (security.Enabled)
                 {
-                    var inputData = new Dictionary<string, object>
+                    var securityTransport = new Coordinator.SecurityCoordinator(security, context, scope.Span);
+                    if (!securityTransport.IsBlocked)
                     {
-                        { AddressesConstants.RequestBody, ObjectExtractor.Extract(bodyDic) },
-                        { AddressesConstants.RequestPathParams, ObjectExtractor.Extract(pathParamsDic) }
-                    };
-                    securityTransport.CheckAndBlock(inputData);
+                        var inputData = new Dictionary<string, object>
+                        {
+                            { AddressesConstants.RequestBody, requestBody },
+                            { AddressesConstants.RequestPathParams, ObjectExtractor.Extract(pathParamsDic) }
+                        };
+                        securityTransport.CheckAndBlock(inputData);
+                    }
                 }
-            }
 
-            if (iastEnabled)
-            {
-                scope?.Span?.Context?.TraceContext?.IastRequestContext?.AddRequestData(context.Request, controllerContext.RouteData.Values);
+                if (iastEnabled)
+                {
+                    scope?.Span?.Context?.TraceContext?.IastRequestContext?.AddRequestData(context.Request, controllerContext.RouteData.Values);
+                    scope?.Span?.Context?.TraceContext?.IastRequestContext?.AddRequestBody(null, requestBody);
```

Comment:
The `AddRequestBody(object body, object? bodyExtracted)` signature specifies that you should _not_ pass `null` to the `body` parameter, but that's what you're doing here. This _should_ be a compiler warning if nullable analysis is enabled.

I think you should _either_ also make `body` an `object?` in `AddRequestBody()` (and handle that it could be `null` inside `AddRequestBody()`), or you should call the `AddExtractedBody()` method here instead (and move the telemetry back where it was). The second one is the best option IMO.

Reply:
I think that I will go for the second option that you mentioned... It makes sense the parameter body to be nullable since it does not need to be used if bodyExtracted is not null. Also, AddExtractedBody might call itself recursively, which would increase more than one the telemetry count. Thanks for the feedback!